File: codeforces/045/F.rb
$a = [
	[],
	[0, 0, 0, 0, 0, 0],
	[0, 1, 3, 6, 0, 0],
	[0, 1, 2, 3, 5, 6]]

def gao(m, n)
	return m > 5 ? 0 : $a[n][m] if n < 4
	s = n / 2
	return 1 if m <= s
	s += (n - 1) / 2
	return 2 if m <= s
	s += n / 2 - 1
	return 3 if m <= s
	s += (n - 1) / 2
	return 4 if m <= s
	t = n / 2 - 1
	return t <= 0 ? 0 : 4 + (m - s + t - 1) / t
end

m, n = gets.split.map(&:to_i)
p gao(m, n) * 2 - 1

__END__
# 	When 	Who 	Problem 	Lang 	Verdict 	Time 	Memory
413397 	Apr 28, 2011 2:47:48 AM 	watashi 	45F - Goats and Wolves 	Ruby 	Accepted 	90 ms 	3800 KB 

File: codeforces/ac/1/177/F.cpp
#include <queue>
#include <cstdio>
#include <vector>
#include <algorithm>
#include <functional>

using namespace std;

bool test(int i, int j) {
    return ((i >> j) & 1) != 0;
}

int set(int i, int j) {
    return i | (1 << j);
}

int main() {
    int n, m, q;
    int a, b, c, d;
    vector<pair<int, pair<int, int>>> e;

    scanf("%d%d%d", &n, &m, &q);
    for (int i = 0; i < m; ++i) {
        scanf("%d%d%d", &a, &b, &c);
        e.push_back({c, {a - 1, b - 1}});
    }
    sort(e.rbegin(), e.rend());

    priority_queue<pair<pair<int, int>, pair<int, int>>> pq;
    pq.push({{0, e.size()}, {0, 0}});
    while (q > 0) {
        d = pq.top().first.first;
        c = pq.top().first.second;
        b = pq.top().second.first;
        a = pq.top().second.second;
        pq.pop();
        if (--q == 0) {
            printf("%d\n", -d);
            break;
        }
        for (int i = c - 1; i >= 0; --i) {
            if (!test(b, e[i].second.first) && !test(a, e[i].second.second)) {
                pq.push({{d - e[i].first, i},
                    {set(b, e[i].second.first), set(a, e[i].second.second)}});
            }
        }
    }

    return 0;
}

/*
# 	When 	Who 	Problem 	Lang 	Verdict 	Time 	Memory
1643545 	Apr 29, 2012 11:34:17 AM 	watashi 	F2 - Script Generation (70 points) 	GNU C++0x 	Accepted 	630 ms 	198400 KB
1643543 	Apr 29, 2012 11:34:10 AM 	watashi 	F1 - Script Generation (30 points) 	GNU C++0x 	Accepted 	30 ms 	1400 KB
*/
